src/Tasks/Microsoft.NET.Build.Tasks/AppHost.cs

```diff
@@ -99,15 +99,15 @@ private static int[] ComputeKMPFailureFunction(byte[] pattern)
         }
 
         // See: https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm
-        private static int KMPSearch(byte[] pattern, byte[] bytes)
+        private static int KMPSearch(byte[] pattern, MemoryMappedViewAccessor accessor)
         {
             int m = 0;
             int i = 0;
             int[] table = ComputeKMPFailureFunction(pattern);
 
-            while (m + i < bytes.Length)
+            while (m + i < accessor.Capacity)
             {
-                if (pattern[i] == bytes[m + i])
+                if (pattern[i] == accessor.ReadByte(m + i))
```

Comment:
This is likely where the slow-down occurs. Did you profile? Each call here is going to be doing some acquire/release ref counting. This is quite a lot of overhead to do for every byte.

You can get a `byte*` from the memory mapped file and drop to unsafe. That is what we do elsewhere where we have used memory mapped files for perf (e.g. System.Reflection.Metadata). Span<byte> would be better, but this code has to run on NET46.

Reply:
It does make a difference. 

Before this change the search (all in memory) need 9461
After the change with accessor.ReadByte need 86683
After use unsafe code, it need 12061

Reply:
I am not very familiar with unsafe in c#, please review with caution

Reply:
> Before this change the search (all in memory) need 9461

9461 what?  ms?  calls?  Does this mean that the new code is slower than what we currently have?

Reply:
in ms. Yes. The old code will load all in memory. Since we need to search for a pattern, that is the most fast way(one round of read is needed anyway). However, the memory consumption is very high in this case since we need to load the whole app host.